pkg/lifecycle/startup/autodetection/autodetection_methods.go

```diff
@@ -0,0 +1,233 @@
+	} else if strings.HasPrefix(method, K8S_INTERNAL_IP) {
+		// K8s InternalIP configured for node is used
+		if k8sNode == nil {
+			log.Error("Cannot use method 'kubernetes-internal-ip' when not running on a Kubernetes cluster")
+			return nil
+		}
+		return autoDetectUsingK8sInternalIP(version, k8sNode, getInterfaces)
+	}
+
+	// The autodetection method is not recognised and is required.  Exit.
+	log.Errorf("Invalid IP autodetection method: %s", method)
+	utils.Terminate()
+	return nil
+}
+
+// autoDetectCIDRFirstFound auto-detects the first valid Network it finds across
+// all interfaces (excluding common known internal interface names).
+func autoDetectCIDRFirstFound(version int, defInterfacesToExclude []string) *cnet.IPNet {
+	incl := []string{}
+
+	iface, cidr, err := FilteredEnumeration(incl, defInterfacesToExclude, nil, version)
+	if err != nil {
+		log.Warnf("Unable to auto-detect an IPv%d address: %s", version, err)
+		return nil
+	}
+
+	log.Infof("Using autodetected IPv%d address on interface %s: %s", version, iface.Name, cidr.String())
+
+	return cidr
+}
+
+// autoDetectCIDRByInterface auto-detects the first valid Network on the interfaces
+// matching the supplied interface regex.
+func autoDetectCIDRByInterface(ifaceRegexes []string, version int) *cnet.IPNet {
+	iface, cidr, err := FilteredEnumeration(ifaceRegexes, nil, nil, version)
+	if err != nil {
+		log.Warnf("Unable to auto-detect an IPv%d address using interface regexes %v: %s", version, ifaceRegexes, err)
+		return nil
+	}
+
+	log.Infof("Using autodetected IPv%d address %s on matching interface %s", version, cidr.String(), iface.Name)
+
+	return cidr
+}
+
+// autoDetectCIDRByCIDR auto-detects the first valid Network on the interfaces
+// matching the supplied cidr.
+func autoDetectCIDRByCIDR(matches []cnet.IPNet, version int) *cnet.IPNet {
+	iface, cidr, err := FilteredEnumeration(nil, nil, matches, version)
+	if err != nil {
+		log.Warnf("Unable to auto-detect an IPv%d address using interface cidr %s: %s", version, matches, err)
+		return nil
+	}
+
+	log.Infof("Using autodetected IPv%d address %s on interface %s matching cidrs %+v", version, cidr.String(), iface.Name, matches)
+
+	return cidr
+}
+
+// autoDetectCIDRByReach auto-detects the IP and Network by setting up a UDP
+// connection to a "reach" address.
+func autoDetectCIDRByReach(dest string, version int) *cnet.IPNet {
+	if cidr, err := ReachDestination(dest, version); err != nil {
+		log.Warnf("Unable to auto-detect IPv%d address by connecting to %s: %s", version, dest, err)
+		return nil
+	} else {
+		log.Infof("Using autodetected IPv%d address %s, detected by connecting to %s", version, cidr.String(), dest)
+		return cidr
+	}
+}
+
+// autoDetectCIDRBySkipInterface auto-detects the first valid Network on the interfaces
+// matching the supplied interface regexes.
+func autoDetectCIDRBySkipInterface(ifaceRegexes []string, version int, defInterfacesToExclude []string) *cnet.IPNet {
```

Comment:
This didn't take interfaces to exclude as an argument before - why do we need this change?

Reply:
yes removed it.